EMAILC-54: IMAP not applying filtering on server side results in larg… (#1093)
mulesoft/docs-connectors: email/1.3/modules/ROOT/pages/email-trigger.adoc

= Trigger a Flow When Receiving a New Email 
:page-aliases: connectors::email/email-trigger.adoc

> Anypoint for Email (Email Connector) provides the *On New Email - IMAP* and *On New Email - POP3* source listeners that trigger a flow by retrieving an email folder from a mailbox for all the unread emails, and generate a new message for each email found. The following examples illustrates how to configure these sources parameters to identify a new email when triggering a flow.

Reviewer: ```suggestion
Anypoint for Email (Email Connector) provides the *On New Email - IMAP* and *On New Email - POP3* source listeners that trigger a flow by retrieving an email folder from a mailbox for all the unread emails. A new message is then generated for each email found. The following examples illustrate how to configure these source parameters to identify a new email when triggering a flow.
```

.Delete after retrieve with On New Email - IMAP operation
image::email-delete-1.png[Delete after retrieve with On New Email - IMAP]

In the XML editor, the configuration `deleteAfterRetrieve="true"` is under the `listener-imap` tag:
[source, xml, linenums]
----
<flow name="OnNewEmail-IMAP">
		<email:listener-imap config-ref="Email_IMAP" watermarkEnabled="false" deleteAfterRetrieve="true">
			<scheduling-strategy >
				<fixed-frequency />
			</scheduling-strategy>
		</email:listener-imap>
	</flow>
----

The following screenshot shows the *Delete after retrieve* configuration for the *On New Email - POP3* source:

.Delete after retrieve with On New Email - POP3 operation
image::email-delete-2.png[Delete after retrieve with On New Email - POP]

In the XML editor, the configuration `deleteAfterRetrieve="true"` is under the `listener-pop3` tag:
[source, xml, linenums]
----
<flow name="OnNewEmail-POP3">
  <email:listener-pop3 config-ref="Email_POP3" deleteAfterRetrieve="true">
    <scheduling-strategy >
      <fixed-frequency />
    </scheduling-strategy>
  </email:listener-pop3>
</flow>
----

> == Configure Enable Watermark Parameter

Reviewer: ```suggestion
== Configure the Enable Watermark Parameter
```

> Email Connector can watermark emails over the IMAP protocol by filtering emails based on their received date. The *On New Email - IMAP* source supports the *Enable Watermark* parameter that enables tp dispatch only messages containing emails received after the last poll executed.

Reviewer: tp dispatch? TP dispatch?  I'm not sure. I don't know what tp dispatch enabled means.

The following example shows how to configure this parameter in Anypoint Studio:

. Select the *On New Email - IMAP* source in the Studio canvas.
. On the source configuration screen, enable the *Enable Watermark* parameter.

The following screenshot shows this parameter configuration for the source:

.Enable Watermark with On New Email - IMAP source
image::email-watermark-1.png[Enable Watermark with On New Email - IMAP]

In the XML editor, the configuration `watermark="true"` is under the `listener-imap` tag:

[source, xml, linenums]
----
<flow name="OnNewEmail-IMAP">
    <email:listener-imap config-ref="Email_IMAP" watermark="true">
        <scheduling-strategy>
            <fixed-frequency/>
        </scheduling-strategy>
    </email:listener-imap>
</flow>
----

== Configure IMAP Matcher and POP3 Matcher Parameters

> Email Connector *On New Email - IMAP* and *On New Email - POP3* sources support *Imap matcher* and *Pop3 matcher* parameters that enables you to custom matcher configurations and provide another way of filtering dispatched messages. Each source matchers can filter emails by dates, addresses, subject, flags and so on.

Reviewer: ```suggestion
Email Connector *On New Email - IMAP* and *On New Email - POP3* sources support *Imap matcher* and *Pop3 matcher* parameters that enable you to configure custom matchers and provide another way of filtering dispatched messages. Each source matcher can filter emails by dates, addresses, subject, flags, and so on.
```

The following example shows how to configure these parameters in Anypoint Studio by dispatching all messages that contain the word `IMPORTANT`:

. Select the *On New Email - IMAP* or *On New Email - POP3* source in the Studio canvas.
. On the source configuration screen, for the *Imap matcher* or *Pop3 matcher* parameters, select either:

* *Edit inline* +

> Set all the inline fields *Seen*, *Answered*, *Deleted*, *Recent* and so on

Reviewer: ```suggestion
Set all the inline fields *Seen*, *Answered*, *Deleted*, *Recent*, and so on.
```

* *Global reference* +

> Create a global reference configuration to set all the inline fields

Reviewer: ```suggestion
Create a global reference configuration to set all the inline fields.
```

[start=3]
. In the *Subject regex* field, set the value as `IMPORTANT`.

The following screenshot shows the *Imap matcher*  edit inline configuration for the *On New Email - IMAP* source:

.Imap matcher with On New Email - IMAP source
image::email-imap-matcher.png[Imap matcher with On New Email - IMAP]

In the XML editor, the configuration `imap-matcher subjectRegex="IMPORTANT"` is under the `listener-imap` tag:

[source, xml, linenums]
----
<flow name="OnNewEmail-IMAP">
    <email:listener-imap config-ref="Email_IMAP">
        <scheduling-strategy>
            <fixed-frequency/>
        </scheduling-strategy>
        <email:imap-matcher subjectRegex="IMPORTANT"/>
    </email:listener-imap>
</flow>
----

== Configure Attachment Naming Strategy Parameter

> Email Connector *On New Email - IMAP*, *On New Email - POP3*, *List - IMAP*, and *List - POP3* operations support the *Attachment naming strategy* parameter that enables you to specify a strategy for naming attachments. You can set the parameter at the global element or override it at the operation level. The parameter has the following options to choose from:

Reviewer: ```suggestion
Email Connector *On New Email - IMAP*, *On New Email - POP3*, *List - IMAP*, and *List - POP3* operations support the *Attachment naming strategy* parameter, which enables you to specify a strategy for naming attachments. You can set the parameter at the global element or override it at the operation level. The parameter has the following options to choose from:
```

  <email:imap-connection host="imap.gmail.com" port="993" user="user@gmail.com" password="mypassword">
  </email:imap-connection>
</email:imap-config>
----

The following example shows how to override the naming strategy at the *On New Email - IMAP* source level:

. Select the *On New Email - IMAP* source in the Studio canvas.
. On the source configuration screen, in the *Advanced* tab, for the *Attachment naming strategy* field select one of the following options:

* *NAME*
* *NAME_HEADERS*
* *NAME_HEADERS_SUBJECT*

The following screenshot shows the *Attachment naming strategy* override at the *On New Email - IMAP level* source:

.Attachment naming strategy override source level
image::email-attachment-2.png[Attachment naming strategy override source level]

In the XML editor, the configuration `attachmentNamingStrategy="NAME"` is under the `listener-imap` tag:

[source, xml, linenums]
----
<flow name="OnNewEmail-IMAP">
		<email:listener-imap  config-ref="Email_IMAP" attachmentNamingStrategy="NAME">
			<scheduling-strategy >
				<fixed-frequency />
			</scheduling-strategy>
		</email:listener-imap>
	</flow>
----

== Configure Enable Remote Search for IMAP Matcher Parameter

> When using *On New Email - IMAP* source and *Imap matcher* parameter, you can specify whether you want the filters to be applied server-side by enabling the *Enable remote search filter* parameter.

Reviewer: ```suggestion
When using the *On New Email - IMAP* source and *Imap matcher* parameters, you can specify whether you want the filters to be applied server-side by enabling the *Enable remote search filter* parameter.
```

> This parameter is disabled by default because some email servers are not fully compliant with `rfc-3501` search terms. Enabling the parameter decreases traffic by reducing the amount of emails brought to client-side for processing.

Reviewer: ```suggestion
This parameter is disabled by default because some email servers are not fully compliant with `rfc-3501` search terms. Enabling the parameter decreases traffic by reducing the number of emails sent to the client-side for processing.
```

> For this feature to be fully functional, the email server should at least

Reviewer 2: @emariotti3 Could you confirm what is missing in the last part of this sentence? Thanks

> For this feature to be fully functional, the email server should at least

Reviewer: I think this sentence got cut off.
